source/utils/algorithms.cpp

```diff
@@ -21,15 +22,15 @@ RoutingTable bfs(std::shared_ptr<IRoutingDevice>& start_device) {
             continue;
         }
         used.insert(device);
-        std::set<std::shared_ptr<ILink>> outlinks = device->get_outlinks();
-        for (std::shared_ptr<ILink> link : outlinks) {
-            auto next_hop = link->get_to();
-            auto curr_device = link->get_from();
+        auto outlinks = device->get_outlinks();
+        for (std::weak_ptr<ILink> link : outlinks) {
+            auto next_hop = link.lock()->get_to();
+            auto curr_device = link.lock()->get_from();
```

Comment:
I think you may use `shared_ptr` here (`get_outlinks` returns set of shared_ptr):
```suggestion
        for (std::shared_ptr<ILink> link : outlinks) {
            auto next_hop = link->get_to();
            auto curr_device = link->get_from();
```